### DISCUSSION TITLE
Combine PNG and HDR Gain Map in HEIC

### DISCUSSION THREAD

**Discussion by andsimakov** [ImageMagick/ImageMagick#3568] [Help]
I'd like to experiment with hand-made HEICs with custom HDR Gain Map to get HDR effect on iPhone 12 Pro Max screen. I have 8-bit PNG photo and manually created PNG HDR Gain Map. How can I create a composed HEIC to combine these two files to get the appropriate result? Please advise CLI options. I saw just adding the Depth Map but I'm in doubt if it will work for HDR Gain Map. Thanks.

**Comment by Michael-Z-Freeman** [comment]
Hi. Where is the adding depth map example and/or documentation ?

**Comment by fmw42** [comment]
What is the appropriate results. Post an input, depth map and an example combined image.

**Reply by Michael-Z-Freeman** [reply]
Hi. Following is the primary image and its depth buffer. Example of final result is here: https://files.michaelzfreeman.org/IMG_0019_A.HEIC

![frame_0000000](https://user-images.githubusercontent.com/951566/218328411-d8389687-39aa-4052-a66b-02be83ff53d1.png)
![frame_0000000_zbuffer](https://user-images.githubusercontent.com/951566/218328425-56fa8c18-15be-4da0-94cf-ffa001af47cc.png)

**Comment by fmw42** [comment]
The HEIC result you post is not for these two images. We need to understand how the two images merge into the final result.

**Reply by Michael-Z-Freeman** [reply]
Actually all I need to know is if ImageMagick can **_ADD_** any auxiliary image to an HEIC image. I also need to know where this is all documented.

If you want to know how the pieces fit together in the HEIC I uploaded then _mp4box -info <HEIC-file>_ (gpac) will show you. _heif-convert_ (of libheif examples) will extract the images.

**Reply by fmw42** [reply]
Sorry can you provide the link to that mp4box -info (gpac)

**Reply by Michael-Z-Freeman** [reply]
No problem: https://gpac.wp.imt.fr/ … mp4box (gpac) is already available in most distribution repositories.

**Comment by snibgo** [comment]
> Following is the primary image and its depth buffer. Example of final result is here: https://files.michaelzfreeman.org/IMG_0019_A.HEIC

I have tried ImageMagick, exiftool and Gimp. None of these systems can see a depth buffer in that file, so I expect those systems can not add a depth buffer to a file.

**Comment by gregbenz** [comment]
Things have come a long way since 2001. There is now great support in most browsers for HDR images. Chrome, Brave, Opera, and Edge support both HDR JPG gain maps as well as standard HDR AVIF (Chrome has support for AVIF gain maps). 

JPG gain maps are clearly the best option at this time as they offer true HDR display, superior SDR display, full backwards compatibility (browsers which don't understand gain maps fall back safely to the SDR version of the image), and are on an ISO-standards track (currently a draft, but the JPG gain map spec published by Adobe / Google is already a de facto standard). Info on gain maps: https://gregbenzphotography.com/hdr-images/jpg-hdr-gain-maps-in-adobe-camera-raw/

Unfortunately, ImageMagick does not support JPG gain maps. A feature request has been submitted here: https://github.com/ImageMagick/ImageMagick/issues/6377

At some point in the future, AVIF gain maps will be the ideal solution as it offers the benefits of gain maps, a much smaller file, and higher quality. Currently, there is no encoder for AVIF gain maps yet, MS Edge does not support any AVIF without enabling a developer flag, and browser support is still limited. However, there is the there are sample images available online and they look amazing in Chrome (Brave also supports if you enable a developer flag brave://flags/#avif-gainmap-hdr-images).
